### PR TITLE
MIDI tests

### PR DESCRIPTION
**Pull request by ed-xmos**
Work for issue: https://xmosjira.atlassian.net/browse/UA-21

### CODE REVIEW

File: tests/xua_unit_tests/CMakeLists.txt [hunk 1]
```diff
@@ -1,103 +1,97 @@
...
+    endif()
+
+
+    # Workaround for xcommon cmake pre-pending CMAKE_CURRENT_LIST_DIR
+    string(REPLACE ${CMAKE_CURRENT_LIST_DIR} "" UNIT_TEST_SOURCE_RELATIVE ${TESTFILE})
+    string(REPLACE ${CMAKE_CURRENT_LIST_DIR} "" RUNNER_FILE_RELATIVE ${RUNNER_FILE})
+
+    set(APP_C_SRCS      ${RUNNER_FILE_RELATIVE}
+                        ${UNIT_TEST_SOURCE_RELATIVE}
+                        ../../../Unity/src/unity.c
```
[the remark this answers is not in the document]
Comment (ed-xmos) [reply tests/xua_unit_tests/CMakeLists.txt:85]: This is a good improvement thanks.

**Review by danielpieczko — APPROVED** [review]

File: tests/test_midi/src/app_midi_simple.xc [hunk 2]
```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 XMOS LIMITED.
+// This Software is subject to the terms of the XMOS Public Licence: Version 1.
+
```
Comment (xross) [inline tests/test_midi/src/app_midi_simple.xc:3]: This comments should be deleted (copy pasta)

File: tests/xua_unit_tests/src/test_midi_parse/test_midi_parse.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+#define RANGE           224
+#define MANUFACTURE_ID  240
+
+#define DATA_RANGE      128
+#define DATA_MASK       (DATA_RANGE - 1)
+
+#define NUM_TESTS_PER_TEST  30
+
+#define CABLE_NUM       2
+#define RANDOM_SEED     6031769
```
Comment (xross) [inline tests/xua_unit_tests/src/test_midi_parse/test_midi_parse.c:25]: you could set this define in pytest to gain extra coverage.

File: tests/xua_unit_tests/src/xua_unit_test_helper.xc [hunk 4]
```diff
@@ -26,3 +29,87 @@ void AudioHwInit()
 {
     ; // nothing
 }
+
+unsigned random(unsigned &x){
```
Comment (xross) [inline tests/xua_unit_tests/src/xua_unit_test_helper.xc:33]: using lib_random might be nicer.

**Review by xross — APPROVED** [review]
I made a few comments but didn't spot any showstoppers. I didn't really spent much time looking a the hid tests, but thanks for sorting those :)

File: tests/test_midi/src/app_midi_simple.xc [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+/* A very simple *example* of a USB audio application (and as such is un-verified for production)
+ *
+ * It uses the main blocks from the lib_xua
+ *
+ * - 2 channels out I2S only
```
Comment (xross) [inline tests/test_midi/src/app_midi_simple.xc:8]: This comments should be deleted (copy pasta)

Comment (ed-xmos) [reply tests/test_midi/src/app_midi_simple.xc:8]: Ah makes more sense!

File: tests/xua_unit_tests/src/xua_unit_test_helper.xc (hunk 4 above)
Comment (ed-xmos) [reply tests/xua_unit_tests/src/xua_unit_test_helper.xc:33]: Hmm lib_random not cmake xcommon compatible yet..

File: tests/test_midi/Makefile
[the remark this answers is not in the document]
Comment (ed-xmos) [reply tests/test_midi/Makefile:1]: Yep - will do this in next PR. Examples still use xcommon.

**Pull request merged**